Gitweb:     http://git.kernel.org/linus/bdddd2963c0264c56f18043f6fa829d3c1d3d1c0
Commit:     bdddd2963c0264c56f18043f6fa829d3c1d3d1c0
Parent:     8592e6486a177a02f048567cb928bc3a1f9a86c3
Author:     Rusty Russell <rusty@rustcorp.com.au>
AuthorDate: Wed Dec 2 14:09:16 2009 +1030
Committer:  Ingo Molnar <mingo@elte.hu>
CommitDate: Wed Dec 2 10:27:16 2009 +0100

    sched: Fix isolcpus boot option
    
    Anton Blanchard wrote:
    
    > We allocate and zero cpu_isolated_map after the isolcpus
    > __setup option has run. This means cpu_isolated_map always
    > ends up empty and if CPUMASK_OFFSTACK is enabled we write to a
    > cpumask that hasn't been allocated.
    
    I introduced this regression in 49557e620339cb13 (sched: Fix
    boot crash by zalloc()ing most of the cpu masks).
    
    Use the bootmem allocator if they set isolcpus=, otherwise
    allocate and zero like normal.
    
    Reported-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
    Cc: peterz@infradead.org
    Cc: Linus Torvalds <torvalds@linux-foundation.org>
    Cc: <stable@kernel.org>
    LKML-Reference: <200912021409.17013.rusty@rustcorp.com.au>
    Signed-off-by: Ingo Molnar <mingo@elte.hu>
    Tested-by: Anton Blanchard <anton@samba.org>
---
 kernel/sched.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/kernel/sched.c b/kernel/sched.c
index 1031cae..4883fee 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -8061,6 +8061,7 @@ static cpumask_var_t cpu_isolated_map;
 /* Setup the mask of cpus configured for isolated domains */
 static int __init isolated_cpu_setup(char *str)
 {
+	alloc_bootmem_cpumask_var(&cpu_isolated_map);
 	cpulist_parse(str, cpu_isolated_map);
 	return 1;
 }
@@ -9609,7 +9610,9 @@ void __init sched_init(void)
 	zalloc_cpumask_var(&nohz.cpu_mask, GFP_NOWAIT);
 	alloc_cpumask_var(&nohz.ilb_grp_nohz_mask, GFP_NOWAIT);
 #endif
-	zalloc_cpumask_var(&cpu_isolated_map, GFP_NOWAIT);
+	/* May be allocated at isolcpus cmdline parse time */
+	if (cpu_isolated_map == NULL)
+		zalloc_cpumask_var(&cpu_isolated_map, GFP_NOWAIT);
 #endif /* SMP */
 
 	perf_event_init();
